### PR TITLE
WIP: multiple section types

### PR DESCRIPTION
**Pull request by Shivansh-yadav13**
ref: https://github.com/casbin/node-casbin/issues/311#issuecomment-1166572177
cherry-pick of [cc58c57](https://github.com/casbin/node-casbin/commit/cc58c57ae4a7c3202c7217497bef8016df93fe26) and [33c784c](https://github.com/casbin/node-casbin/commit/33c784cbdc0d650ba75b8177b69f41dc0ca11fae)

### CODE REVIEW

**Comment by hsluoyz** [comment]
@Shivansh-yadav13 can you post a video containing the following pic?

image

**Comment by hsluoyz** [comment]
@Shivansh-yadav13 OK, I found it in the system, perhaps the git bot is broken.

@nodece @Zxilly plz review

File: src/coreEnforcer.ts [hunk 1]
```diff
@@ -551,7 +559,11 @@ export class CoreEnforcer {
    * @return whether to allow the request.
    */
   public enforceSync(...rvals: any[]): boolean {
-    return generatorRunSync(this.privateEnforce(false, false, ...rvals));
+    if (rvals[0] instanceof EnforceContext) {
```
Comment (Zxilly) [inline src/coreEnforcer.ts:562]: should check length of `rvals` before pick element from it.

File: src/coreEnforcer.ts [hunk 2]
```diff
@@ -565,7 +577,11 @@ export class CoreEnforcer {
    * @return whether to allow the request and the reason rule.
    */
   public enforceExSync(...rvals: any[]): [boolean, string[]] {
-    return generatorRunSync(this.privateEnforce(false, true, ...rvals));
+    if (rvals[0] instanceof EnforceContext) {
```
Comment (Zxilly) [inline src/coreEnforcer.ts:580]: same

File: src/coreEnforcer.ts [hunk 3]
```diff
@@ -584,7 +600,11 @@ export class CoreEnforcer {
    * @return whether to allow the request.
    */
   public async enforce(...rvals: any[]): Promise<boolean> {
-    return generatorRunAsync(this.privateEnforce(true, false, ...rvals));
+    if (rvals[0] instanceof EnforceContext) {
```
Comment (Zxilly) [inline src/coreEnforcer.ts:603]: same

File: src/coreEnforcer.ts [hunk 4]
```diff
@@ -596,7 +616,11 @@ export class CoreEnforcer {
    * @return whether to allow the request and the reason rule.
    */
   public async enforceEx(...rvals: any[]): Promise<[boolean, string[]]> {
-    return generatorRunAsync(this.privateEnforce(true, true, ...rvals));
+    if (rvals[0] instanceof EnforceContext) {
```
Comment (Zxilly) [inline src/coreEnforcer.ts:619]: same

File: src/persist/index.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
 export * from './adapter';
 export * from './fileAdapter';
-export * from './stringAdapter';
+export * from './memoryAdapter';
```
Comment (Zxilly) [inline src/persist/index.ts:3]: this will lead to breaking change. Please also export memory adapter.

File: src/persist/stringAdapter.ts [hunk 6]
```diff
@@ -1,65 +0,0 @@
-import { Adapter } from './adapter';
```
Comment (Zxilly) [inline src/persist/stringAdapter.ts:1]: breaking change

**Review by Zxilly — CHANGES_REQUESTED** [review]
Please don't modify code of `StringAdapter`. We should not introduce breaking change as for now.

File: test/utils.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import { readFileSync } from 'fs';
```
Comment (nodece) [inline test/utils.ts:1]: Remove this file.

File: src/persist/memoryAdapter.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
+import { Adapter } from './adapter';
```
Comment (nodece) [inline src/persist/memoryAdapter.ts:1]: Revert the `StringAdapter.ts`, and remove this file.

File: test/enforcer.test.ts [hunk 9]
```diff
@@ -697,3 +691,71 @@ test('TestEnforceExWithPriorityModel', async () => {
...
 });
+
+test('TestEnforce Multiple policies config', async () => {
+  const m = newModel();
+  m.addDef('r', 'r2', 'sub, obj, act');
+  m.addDef('p', 'p2', 'sub, obj, act');
+  m.addDef('g', 'g', '_, _');
+  m.addDef('e', 'e2', 'some(where (p.eft == allow))');
+  m.addDef('m', 'm2', 'g(r2.sub, p2.sub) && r2.obj == p2.obj && r2.act == p2.act');
+  const a = getStringAdapter('examples/mulitple_policy.csv');
```
Comment (nodece) [inline test/enforcer.test.ts:702]:
```
const policy = readFileSync('examples/mulitple_policy.csv').toString();
const a = new StringAdapter(policy);
```

File: test/enforcer.test.ts [hunk 10]
```diff
@@ -697,3 +691,71 @@ test('TestEnforceExWithPriorityModel', async () => {
...
+});
+
+test('new EnforceContext config', async () => {
+  const m = newModel();
+  m.addDef('r', 'r2', 'sub, obj, act');
+  m.addDef('p', 'p2', 'sub, obj, act');
+  m.addDef('g', 'g', '_, _');
+  m.addDef('e', 'e2', 'some(where (p.eft == allow))');
+  m.addDef('m', 'm2', 'g(r2.sub, p2.sub) && r2.obj == p2.obj && r2.act == p2.act');
+  const a = getStringAdapter('examples/mulitple_policy.csv');
```
Comment (nodece) [inline test/enforcer.test.ts:719]:
```
const policy = readFileSync('examples/mulitple_policy.csv').toString();
const a = new StringAdapter(policy);
```

File: test/enforcer.test.ts [hunk 11]
```diff
@@ -697,3 +691,71 @@ test('TestEnforceExWithPriorityModel', async () => {
...
+});
+
+test('TestEnforceEX Multiple policies config', async () => {
+  const m = newModel();
+  m.addDef('r', 'r2', 'sub, obj, act');
+  m.addDef('p', 'p2', 'sub, obj, act');
+  m.addDef('g', 'g', '_, _');
+  m.addDef('e', 'e2', 'some(where (p.eft == allow))');
+  m.addDef('m', 'm2', 'g(r2.sub, p2.sub) && r2.obj == p2.obj && r2.act == p2.act');
+  const a = getStringAdapter('examples/mulitple_policy.csv');
```
Comment (nodece) [inline test/enforcer.test.ts:736]:
```
const policy = readFileSync('examples/mulitple_policy.csv').toString();
const a = new StringAdapter(policy);
```

File: test/enforcer.test.ts [hunk 12]
```diff
@@ -697,3 +691,71 @@ test('TestEnforceExWithPriorityModel', async () => {
...
+});
+
+test('new EnforceContextEX config', async () => {
+  const m = newModel();
+  m.addDef('r', 'r2', 'sub, obj, act');
+  m.addDef('p', 'p2', 'sub, obj, act');
+  m.addDef('g', 'g', '_, _');
+  m.addDef('e', 'e2', 'some(where (p.eft == allow))');
+  m.addDef('m', 'm2', 'g(r2.sub, p2.sub) && r2.obj == p2.obj && r2.act == p2.act');
+  const a = getStringAdapter('examples/mulitple_policy.csv');
```
Comment (nodece) [inline test/enforcer.test.ts:753]:
```
const policy = readFileSync('examples/mulitple_policy.csv').toString();
const a = new StringAdapter(policy);
```

File: src/persist/index.ts (hunk 5 above)
Comment (nodece) [reply src/persist/index.ts:3]: Revert the `stringAdapter`.

**Review by nodece — CHANGES_REQUESTED** [review]

**Comment by nodece** [comment]
@Shivansh-yadav13 Any updates?

**Comment by nodece** [comment]
@Shivansh-yadav13 You look like push an incorrect commit, please check this PR.

**Comment by VictrumMaximum** [comment]
Hello, is there any update on this feature? I see there was some progress, but it got overwritten by a git mistake... I am considering helping out, but I also don't want to start from scratch if I don't need to.

**Comment by nodece** [comment]
Closed by https://github.com/casbin/node-casbin/pull/442.

**Pull request closed without merging**